### PR TITLE
Cassandra 16259 (trunk)

### PR DESCRIPTION
**Pull request by blerer**
1. Fix the histogram merge of the table metrics
2. Fix the counting of cells per partition

### CODE REVIEW

File: src/java/org/apache/cassandra/db/transform/BaseIterator.java [hunk 1]
```diff
@@ -83,14 +87,31 @@
 
     public final void close()
     {
+        // If close has already been called we want to ignore other calls
+        if (closed)
```
Comment (bereng) [inline src/java/org/apache/cassandra/db/transform/BaseIterator.java:91]: Unless I am missing sthg: I'd wrap everything in a try-finally block and toggle `closed` to true only there. That is when it is indeed closed and not 'closing' as it is at this point. I can see the code _tries_ to catch/handle all failures but `runOnClose()` is abstract to be implemented in any way.

That would prevent a failed close operation to toggle the `closed` variable erroneously.

Comment (blerer) [reply src/java/org/apache/cassandra/db/transform/BaseIterator.java:91]: If the iterator has been registered in multiples `try-with-resources`  (the root of the problem we are trying to solve) and a bad implementation of `runOnClose` fire an Exception, the next `try-with-resources` will hit the same problem and will hide the original one. No matter how the execution of `close` end up, I do not think that there are any scenarios in which you want to retry closing the iterator. Which is why the patch toggle `closed` as early as possible.

**Comment by bereng** [comment]
You can have someone assume an idempotent `close()` and for some weird reason it gets called multiple times just as you found. Also in between those calls the env might have changed enough later calls succeed.

In any case I don't have a strong position here, just trying to future proof it. Feel free to ignore it.

File: test/unit/org/apache/cassandra/db/ColumnFamilyMetricTest.java [hunk 2]
```diff
@@ -140,6 +133,101 @@ public void testStartupRaceConditionOnMetricListeners()
...
+
+        long[] empty = new long[0];
+
+        result = TableMetrics.addHistogram(sums, empty);
+        assertTrue(result == sums); // Check that we did not create a new array
+        assertArrayEquals(new long[]{6, 8, 7}, result);
+
+        long[] greater = new long[] {4, 3, 2, 1};
+        result = TableMetrics.addHistogram(sums, greater);
+        assertFalse(result == sums); // Check that we did not create a new array
```
Comment (jasonstack) [inline test/unit/org/apache/cassandra/db/ColumnFamilyMetricTest.java:207]:
```suggestion
        assertFalse(result == sums); // Check that we did create a new array
```

**Review by jasonstack — APPROVED** [review]

**Comment by blerer** [comment]
Merged manually.

**Pull request closed without merging**